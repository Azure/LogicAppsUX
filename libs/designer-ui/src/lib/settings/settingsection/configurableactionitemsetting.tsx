import { Checkbox, FontSizes, IconButton } from '@fluentui/react';
import type { IButtonStyles, IIconProps } from '@fluentui/react';
import { useBoolean } from '@fluentui/react-hooks';
import { isNullOrUndefined } from '@microsoft-logic-apps/utils';
import React, { useEffect, useState } from 'react';
import { useIntl } from 'react-intl';

export interface ConfigurableMultiSelectItemSectionProps {
  configurableActionItems: ConfigurableAction[];
  configuredActionItems: ConfigurableAction[];
  isReadOnly: boolean;
}

export interface ConfigurableAction {
  id: string;
  title: string;
  icon?: string;
  statuses: Record<string, boolean>;
}

type StatusChangeHandler = (statusKey: string, checked?: boolean) => void;

export const ConfigurableMultiSelectItemSection: React.FC<ConfigurableMultiSelectItemSectionProps> = (props): JSX.Element => {
  const canDelete = (item: ConfigurableAction, actions: ConfigurableAction[]): boolean => actions.length > 1;
  const { configurableActionItems, configuredActionItems, isReadOnly } = props;
  const defaultStatuses: Record<string, boolean> = {
    'Option 1': true,
    'Option 2': false,
    'Option 3': false,
    'Option 4': false,
  };
  const [items, setItems] = useState(configuredActionItems);

  const onStatusChange = (itemId: string, statusKey: string, checked?: boolean): void => {
    if (isNullOrUndefined(checked)) {
      return;
    }
    const itemIndex = items.indexOf(items.filter((item) => item.id === itemId)[0]);
    const itemsCopy = [...items];
    const currentItem = items[itemIndex];
    const modifiedItem: ConfigurableAction = { ...currentItem, statuses: { ...currentItem.statuses, [statusKey]: checked } };
    itemsCopy.splice(itemIndex, 1, modifiedItem);
    setItems(itemsCopy);
  };
  // const handleActionAddition = (id: string): void => {
  //   // construct action object via id -> add new action to state items object
  //   let newAction: ConfigurableAction = {
  //     id,
  //     title: 'Random Action',
  //     icon: '',
  //     statuses: defaultStatuses,
  //   };
  //   setItems([...items, newAction]);
  // };
  const handleActionDelete = (id: string): void => {
    const indexOfItemToDelete = items.indexOf(items.filter((item) => item.id === id)[0]);
    const itemsCopy = [...items];
    itemsCopy.splice(indexOfItemToDelete, 1);
    setItems(itemsCopy);
  };
  return (
    <>
      {items.map((configuredAction) => {
        return (
          <ActionItemConfiguration
            key={configuredAction.id}
            onDelete={handleActionDelete}
            action={configuredAction}
            canDelete={canDelete(configuredAction, items)}
            onStatusChange={(statusKey: string, checked?: boolean) => onStatusChange(configuredAction.id, statusKey, checked)}
            isReadOnly={isReadOnly}
          />
        );
      })}
      <div className="'msla-run-after-action-add">
        {/* <AddActionItem>
            configurableItems={configurableActionItems}
            preconfiguredItems={configuredActionItems}
            onActionAdd={handleActionAddition}
        </AddActionItem> */}
      </div>
    </>
  );
};

interface ActionHeaderProps {
  id: string;
  icon?: string;
  title: string;
  customRender?(): JSX.Element | null;
}

interface ActionItemConfigurationProps {
  action: ConfigurableAction;
  canDelete: boolean;
  onDelete: DeleteActionHandler;
  onStatusChange: StatusChangeHandler;
  isReadOnly: boolean;
}

const ActionItemConfiguration: React.FC<ActionItemConfigurationProps> = (props): JSX.Element => {
  const {
    action: { statuses, id, icon, title },
    canDelete,
    onDelete,
    onStatusChange,
    isReadOnly,
  } = props;
  const [expanded, toggleExpand] = useBoolean(!canDelete);
  const toggleActionExpand = (_: React.MouseEvent<HTMLElement>): void => {
    toggleExpand.toggle();
  };
  const ActionHeader: React.FC<ActionHeaderProps> = (props): JSX.Element => {
    const { customRender } = props;
    const iconProps: IIconProps = { iconName: expanded ? 'ChevronDownMed' : 'ChevronRightMed' };
    const chevronStyles: IButtonStyles = {
      icon: {
        color: '#8a8886',
        fontSize: FontSizes.small,
      },
    };
    const intl = useIntl();
    const expandOrCollapse = intl.formatMessage({
      defaultMessage: 'Expand or Collapse',
      description: 'description to toggle expansion of section',
    });
    return (
      <div className="msla-run-after-edge-header">
        <div className="msla-run-after-edge-header-contents-container" onClick={toggleActionExpand}>
          <div>
            <div className="msla-run-after-edge-header-contents">
              <IconButton ariaLabel={expandOrCollapse} className="msla-run-after-icon" iconProps={iconProps} styles={chevronStyles} />
              <div className="msla-run-after-edge-header-logo">
                <img
                  alt=""
                  className="msla-run-after-logo-image"
                  role="presentation"
                  src={icon ?? 'https://reactnativecode.com/wp-content/uploads/2018/02/Default_Image_Thumbnail.png'}
                />
              </div>
              <div className="msla-run-after-edge-header-text">{title}</div>
            </div>
            {customRender}
            <DeleteButton id={id} visible={canDelete && !isReadOnly} onDelete={onDelete} />
          </div>
        </div>
      </div>
    );
  };
  const actionHeaderProps: ActionHeaderProps = {
    id,
    title,
    icon,
  };

  return (
    <>
      <ActionHeader {...actionHeaderProps} />
      {expanded ? <ActionStatuses id={id} statuses={statuses} onStatusChange={onStatusChange} /> : null}
    </>
  );
};

type DeleteActionHandler = (id: string) => void;

interface DeleteButtonProps {
  id: string;
  visible: boolean;
  onDelete: DeleteActionHandler;
}

const DeleteButton: React.FC<DeleteButtonProps> = (props): JSX.Element | null => {
  const { onDelete, id, visible } = props;
  const intl = useIntl();
  const deleteAriaLabel = intl.formatMessage({
    defaultMessage: 'Delete',
    description: 'aria label for Delete icon',
  });
  const deleteIconProp: IIconProps = {
    iconName: 'Delete',
  };
  const deleteIconStyles: IButtonStyles = {
    icon: {
      color: '#0078d4',
      fontSize: 16,
    },
  };
  const handleDelete = (e: React.MouseEvent<HTMLElement>): void => {
    onDelete(id);
  };

  if (!visible) {
    return null;
  }

  return (
    <div className="msla-run-after-delete-icon">
      <IconButton ariaLabel={deleteAriaLabel} iconProps={deleteIconProp} styles={deleteIconStyles} onClick={handleDelete} />
    </div>
  );
};

export interface ActionStatusesProps {
  id: string;
  statuses: Record<string, boolean>;
  onStatusChange: StatusChangeHandler;
  onRenderLabel?(status: string, label: string): JSX.Element | null;
}

const ActionStatuses: React.FC<ActionStatusesProps> = (props): JSX.Element => {
  const { statuses, onStatusChange, onRenderLabel } = props;
  const intl = useIntl();
  const option1Label = intl.formatMessage({
    defaultMessage: 'Option 1',
    description: 'label for first status checkbox',
  });
  const option2Label = intl.formatMessage({
    defaultMessage: 'Option 2',
    description: 'label for second status checkbox',
  });
  const option3Label = intl.formatMessage({
    defaultMessage: 'Option 3',
    description: 'label for third status checkbox',
  });
  const option4Label = intl.formatMessage({
    defaultMessage: 'Option 4',
    description: 'label for fourth status checkbox',
  });
  return (
    <div className="msla-run-after-statuses">
      <div className="msla-run-after-status-checkbox">
        <Checkbox
          checked={statuses[option1Label]}
          label={option1Label}
          onChange={(_, checked) => onStatusChange?.(option1Label, checked)}
        />
      </div>
      <div className="msla-run-after-status-checkbox">
        <Checkbox
          checked={statuses[option2Label]}
          label={option2Label}
          onChange={(_, checked) => onStatusChange?.(option2Label, checked)}
        />
      </div>
      <div className="msla-run-after-status-checkbox">
        <Checkbox
          checked={statuses[option3Label]}
          label={option3Label}
          onChange={(_, checked) => onStatusChange?.(option3Label, checked)}
        />
      </div>
      <div className="msla-run-after-status-checkbox">
        <Checkbox
          checked={statuses[option4Label]}
          label={option4Label}
          onChange={(_, checked) => onStatusChange?.(option4Label, checked)}
        />
      </div>
    </div>
  );
};
