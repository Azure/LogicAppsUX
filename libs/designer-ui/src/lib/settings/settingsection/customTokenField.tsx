import type { ValueSegment } from '../../editor';
import type { ChangeHandler } from '../../editor/base';
import type { TokenFieldProps } from './settingTokenField';
import { TokenField } from './settingTokenField';
import type {
  ChangeHandler as CustomEditorChangeHandler,
  ICustomEditorOptions,
  IEditorProps,
  IRenderDefaultEditorParams,
} from '@microsoft/designer-client-services-logic-apps';
import { equals } from '@microsoft/logic-apps-shared';
import { useCallback } from 'react';

export type CustomTokenFieldProps = Omit<TokenFieldProps, 'editor' | 'editorOptions'> & ICustomEditorAndOptions;

export const CustomTokenField = (props: CustomTokenFieldProps) => {
  const { EditorComponent, editor, editorOptions = {} }: ICustomEditorOptions = props.editorOptions;

  const renderDefaultEditor = useRenderDefaultEditor(props);

  const customEditorProps: IEditorProps = {
    editor,
    editorOptions,
    value: props.value,
    onValueChange: props.onValueChange as CustomEditorChangeHandler,
    renderDefaultEditor,
    disabled: props.readOnly,
  };
  return <EditorComponent {...customEditorProps} />;
};

function useRenderDefaultEditor(tokenFieldProps: Omit<TokenFieldProps, 'editor' | 'editorOptions' | 'value' | 'onValueChange'>) {
  return useCallback(
    (props: IRenderDefaultEditorParams) => {
      return (
        <TokenField
          {...tokenFieldProps}
          editor={props.editor}
          editorOptions={props.editorOptions}
          value={props.value as ValueSegment[]}
          onValueChange={props.onValueChange as ChangeHandler}
        />
      );
    },
    [tokenFieldProps]
  );
}

// Overridden custom editors are identified by this name internally.
const customEditorName = 'internal-custom-editor';

export type ICustomEditorAndOptions = { editor: typeof customEditorName; editorOptions: ICustomEditorOptions };

export const isCustomEditor = (props: { editor?: string | undefined; editorOptions?: unknown }): props is ICustomEditorAndOptions => {
  const { editor, editorOptions } = props;
  return (
    equals(editor, customEditorName) &&
    typeof editorOptions == 'object' &&
    !!editorOptions &&
    typeof (editorOptions as { EditorComponent: unknown }).EditorComponent === 'function'
  );
};

export const toCustomEditorAndOptions = (options: ICustomEditorOptions) => ({
  editor: customEditorName,
  editorOptions: options,
});
