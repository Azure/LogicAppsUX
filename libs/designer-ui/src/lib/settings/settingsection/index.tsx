export { MultiSelectSetting } from './settingmultiselect';
export type { MultiSelectSettingProps } from './settingmultiselect';
export { MultiAddExpressionEditor, ExpressionsEditor, Expressions, Expression } from './settingexpressioneditor';
export type {
  MultiAddExpressionEditorProps,
  ExpressionsEditorProps,
  ExpressionsProps,
  ExpressionProps,
  ExpressionChangeHandler,
} from './settingexpressioneditor';
export { ReactiveToggle } from './settingreactiveinput';
export type { ReactiveToggleProps } from './settingreactiveinput';
export { CustomValueSlider } from './settingslider';
export type { CustomValueSliderProps } from './settingslider';
export { SettingTextField } from './settingtextfield';
export type { SettingTextFieldProps } from './settingtextfield';
export { SettingToggle } from './settingtoggle';
export type { SettingToggleProps, ToggleChangeHandler } from './settingtoggle';
export { SettingLabel, getSettingLabel } from './settinglabel';
export { SettingDictionary } from './settingdictionary';
export type { SettingDictionaryProps } from './settingdictionary';
export { SettingTokenField } from './settingTokenField';
export type { SettingTokenFieldProps as SettingTokenTextFieldProps } from './settingTokenField';
export { SettingDropdown } from './settingdropdown';
export type { SettingDropdownProps, DropdownSelectionChangeHandler } from './settingdropdown';
export { toCustomEditorAndOptions, isCustomEditor } from './customTokenField';

export interface SettingProps {
  readOnly?: boolean;
  ariaLabel?: string;
  customLabel?: JSX.Element;
  nodeTitle?: string;
}
