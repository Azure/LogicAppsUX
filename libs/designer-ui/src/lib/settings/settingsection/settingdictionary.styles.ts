import { makeStyles } from '@fluentui/react-components';

export const useStyles = makeStyles({
  root: {
    width: '100%',
  },
  textFieldWrapper: {
    display: 'inline-flex',
    width: '100%',
  },
  textField: {
    width: '100%',
  },
  dictionaryContainer: {
    width: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  dictionaryWrapper: {
    width: '100%',
    maxWidth: 'none',
  },
});
