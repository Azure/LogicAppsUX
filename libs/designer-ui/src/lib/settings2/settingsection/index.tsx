export interface SettingSectionComponentProps extends Record<string, any> {
  id: string;
  title: string;
  expanded: boolean;
  renderContent: React.FC<any>;
  isInverted: boolean;
  isReadOnly: boolean;
}
