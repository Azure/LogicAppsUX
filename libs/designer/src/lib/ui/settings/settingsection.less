.msla-conditional-parameter-dropdown {
  margin: 24px auto 0 auto;
}

.msla-conditional-parameter-dropdown-search {
  margin: 10px;
}
