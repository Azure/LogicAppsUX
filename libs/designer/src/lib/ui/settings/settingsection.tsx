import type { HeaderClickHandler, SettingSectionName } from '.';
import constants from '../../common/constants';
import { useReadOnly } from '../../core/state/designerOptions/designerOptionsSelectors';
import { updateParameterConditionalVisibility } from '../../core/state/operation/operationMetadataSlice';
import { useSelectedNodeId } from '../../core/state/panel/panelSelectors';
import type { RunAfterProps } from './sections/runafterconfiguration';
import { RunAfter } from './sections/runafterconfiguration';
import { CustomizableMessageBar } from './validation/errorbar';
import type { ValidationError } from './validation/validation';
import { ValidationErrorType } from './validation/validation';
import type { IDropdownOption } from '@fluentui/react';
import { Separator, useTheme, Icon } from '@fluentui/react';
import { Button, Divider, Tooltip } from '@fluentui/react-components';
import { bundleIcon, Dismiss24Filled, Dismiss24Regular } from '@fluentui/react-icons';
import { MessageBarType } from '@fluentui/react/lib/MessageBar';
import {
  isHighContrastBlack,
  MultiSelectSetting,
  MultiAddExpressionEditor,
  ExpressionsEditor,
  Expressions,
  Expression,
  ReactiveToggle,
  CustomValueSlider,
  SettingTextField,
  SettingToggle,
  SettingDictionary,
  SettingTokenField,
  SettingDropdown,
  SearchableDropdownWithAddAll,
} from '@microsoft/designer-ui';
import type {
  MultiSelectSettingProps,
  MultiAddExpressionEditorProps,
  ExpressionsEditorProps,
  ExpressionsProps,
  ExpressionProps,
  ReactiveToggleProps,
  CustomValueSliderProps,
  SettingTextFieldProps,
  SettingTokenTextFieldProps,
  SettingToggleProps,
  SettingDictionaryProps,
  SettingDropdownProps,
  ChangeState,
} from '@microsoft/designer-ui';
import type { FC } from 'react';
import { useMemo, useState } from 'react';
import { useIntl } from 'react-intl';
import { useDispatch } from 'react-redux';

const ClearIcon = bundleIcon(Dismiss24Filled, Dismiss24Regular);

type SettingBase = {
  visible?: boolean;
};

export type Settings = SettingBase &
  (
    | {
        settingType: 'MultiSelectSetting';
        settingProp: MultiSelectSettingProps;
      }
    | {
        settingType: 'MultiAddExpressionEditor';
        settingProp: MultiAddExpressionEditorProps;
      }
    | {
        settingType: 'ExpressionsEditor';
        settingProp: ExpressionsEditorProps;
      }
    | {
        settingType: 'Expressions';
        settingProp: ExpressionsProps;
      }
    | {
        settingType: 'Expression';
        settingProp: ExpressionProps;
      }
    | {
        settingType: 'ReactiveToggle';
        settingProp: ReactiveToggleProps;
      }
    | {
        settingType: 'CustomValueSlider';
        settingProp: CustomValueSliderProps;
      }
    | {
        settingType: 'SettingTextField';
        settingProp: SettingTextFieldProps;
      }
    | {
        settingType: 'SettingToggle';
        settingProp: SettingToggleProps;
      }
    | {
        settingType: 'SettingDictionary';
        settingProp: SettingDictionaryProps;
      }
    | {
        settingType: 'SettingTokenField';
        settingProp: SettingTokenTextFieldProps;
      }
    | {
        settingType: 'RunAfter';
        settingProp: RunAfterProps;
      }
    | {
        settingType: 'SettingDropdown';
        settingProp: SettingDropdownProps;
      }
  );

type WarningDismissHandler = (key?: string, message?: string) => void;
export interface SettingsSectionProps {
  id?: string;
  title?: string;
  sectionName?: string;
  showHeading?: boolean;
  showSeparator?: boolean;
  expanded?: boolean;
  settings: Settings[];
  isReadOnly?: boolean;
  onHeaderClick?: HeaderClickHandler;
  validationErrors?: ValidationError[];
  onDismiss?: WarningDismissHandler;
}

export const SettingsSection: FC<SettingsSectionProps> = ({
  id,
  title = 'Settings',
  sectionName,
  showHeading = true,
  showSeparator = true,
  expanded,
  isReadOnly,
  settings,
  onHeaderClick,
  validationErrors,
  onDismiss,
}) => {
  const theme = useTheme();
  const isInverted = isHighContrastBlack() || theme.isInverted;

  const separatorStyles = {
    root: { color: isInverted ? constants.Settings.SETTING_SEPARATOR_COLOR_DARK : constants.Settings.SETTING_SEPARATOR_COLOR_LIGHT },
  };
  const intl = useIntl();
  const expandAriaLabel = intl.formatMessage({
    defaultMessage: 'Expand',
    description: 'An accessible label for expand toggle icon',
  });
  const collapseAriaLabel = intl.formatMessage({
    defaultMessage: 'Collapse',
    description: 'An accessible label for collapse toggle icon',
  });
  const internalSettings = (
    <>
      {expanded || !showHeading ? <Setting id={id} isReadOnly={isReadOnly} settings={settings} /> : null}
      {expanded
        ? (validationErrors ?? []).map(({ key: errorKey, errorType, message }, i) => {
            return (
              <CustomizableMessageBar
                key={i}
                type={matchErrorTypeToMessageBar(errorType)}
                message={message}
                onWarningDismiss={onDismiss ? () => onDismiss?.(errorKey) : undefined}
              />
            );
          })
        : null}
      {showSeparator ? <Separator className="msla-setting-section-separator" styles={separatorStyles} /> : null}
    </>
  );
  if (!showHeading) {
    return internalSettings;
  }
  const handleSectionClick = (sectionName?: SettingSectionName): void => {
    if (onHeaderClick && sectionName) {
      onHeaderClick(sectionName);
    }
  };

  return (
    <div className="msla-setting-section">
      <div className="msla-setting-section-content">
        <button className="msla-setting-section-header" onClick={() => handleSectionClick(sectionName as SettingSectionName | undefined)}>
          <Icon
            className="msla-setting-section-header-icon"
            aria-label={expanded ? `${collapseAriaLabel} ${title}` : `${expandAriaLabel} ${title}`}
            iconName={expanded ? 'ChevronDownMed' : 'ChevronRightMed'}
            styles={{ root: { fontSize: 14, color: isInverted ? 'white' : constants.Settings.CHEVRON_ROOT_COLOR_LIGHT } }}
          />
          <div className="msla-setting-section-header-text">{title}</div>
        </button>
        {internalSettings}
      </div>
    </div>
  );
};

const Setting = ({ id, settings, isReadOnly }: { id?: string; settings: Settings[]; isReadOnly?: boolean }): JSX.Element => {
  const intl = useIntl();
  const dispatch = useDispatch();
  const nodeId = useSelectedNodeId();
  const readOnly = useReadOnly();
  const [hideErrorMessage, setHideErrorMessage] = useState<boolean[]>(new Array(settings.length).fill(false));

  const updateHideErrorMessage = (index: number, b: boolean) => {
    setHideErrorMessage([...hideErrorMessage.slice(0, index), b, ...hideErrorMessage.slice(index + 1)]);
  };

  const allConditionalSettings = useMemo(
    () => settings.filter((setting) => (setting.settingProp as any).conditionalVisibility !== undefined),
    [settings]
  );

  const conditionallyInvisibleSettings = useMemo(
    () => settings.filter((setting) => (setting.settingProp as any).conditionalVisibility === false),
    [settings]
  );

  const addNewParamText = intl.formatMessage(
    {
      defaultMessage: 'Showing {countShowing} of {countTotal}',
      description: 'Placeholder text for the number of advanced parameters showing',
    },
    {
      countShowing: allConditionalSettings.length - conditionallyInvisibleSettings.length,
      countTotal: allConditionalSettings.length,
    }
  );

  const addAllButtonText = intl.formatMessage({
    defaultMessage: 'Show all',
    description: 'Button text to add all advanced parameters',
  });
  const removeAllButtonText = intl.formatMessage({
    defaultMessage: 'Clear all',
    description: 'Button text to clear all advanced parameters',
  });
  const addAllButtonTooltip = intl.formatMessage({
    defaultMessage: 'Show all advanced parameters',
    description: 'Button tooltip to add all advanced parameters',
  });
  const removeAllButtonTooltip = intl.formatMessage({
    defaultMessage: 'Remove and clear all advanced parameters and their values',
    description: 'Button tooltip to remove all advanced parameters',
  });
  const advancedParametersLabel = intl.formatMessage({
    defaultMessage: 'Advanced parameters',
    description: 'The label for advanced parameters',
  });

  const renderSetting = (setting: Settings, i: number) => {
    const { settingType, settingProp, visible = true } = setting;
    const { id: parameterId, conditionalVisibility, readOnly, validationErrors } = settingProp as any;
    if (!readOnly) settingProp.readOnly = isReadOnly;
    const errorMessage = validationErrors?.reduce((acc: string, message: any) => acc + message + ' ', '');

    const getClassName = (): string =>
      settingType === 'RunAfter'
        ? 'msla-setting-section-run-after-setting'
        : settingType === 'MultiAddExpressionEditor'
        ? 'msla-setting-section-expression-field'
        : 'msla-setting-section-setting';
    const renderSetting = (): JSX.Element | null => {
      switch (settingType) {
        case 'MultiSelectSetting':
          return <MultiSelectSetting {...settingProp} />;
        case 'MultiAddExpressionEditor':
          return <MultiAddExpressionEditor {...settingProp} />;
        case 'ExpressionsEditor':
          return <ExpressionsEditor {...settingProp} />;
        case 'Expressions':
          return <Expressions {...settingProp} />;
        case 'Expression':
          return <Expression {...settingProp} />;
        case 'ReactiveToggle':
          return <ReactiveToggle {...settingProp} />;
        case 'CustomValueSlider':
          return <CustomValueSlider {...settingProp} />;
        case 'SettingTextField':
          return <SettingTextField {...settingProp} />;
        case 'SettingToggle':
          return <SettingToggle {...settingProp} />;
        case 'SettingDictionary':
          return <SettingDictionary {...settingProp} />;
        case 'SettingTokenField':
          return (
            <SettingTokenField
              {...{
                hideValidationErrors: (newState: ChangeState) => {
                  updateHideErrorMessage(i, newState.viewModel.hideErrorMessage);
                },
                ...settingProp,
              }}
            />
          );
        case 'RunAfter':
          return <RunAfter {...settingProp} />;
        case 'SettingDropdown':
          return <SettingDropdown {...settingProp} />;
        default:
          return null;
      }
    };

    const removeParamCallback = () => {
      dispatch(updateParameterConditionalVisibility({ nodeId, groupId: id ?? '', parameterId, value: false }));
    };

    const removeParamTooltip = intl.formatMessage(
      {
        defaultMessage: "Remove parameter ''{parameterName}'' and its value",
        description:
          'Tooltip for button to remove parameter. Do not remove the double single quotes around the display name, as it is needed to wrap the placeholder text.',
      },
      { parameterName: (settingProp as any).label }
    );

    return visible && conditionalVisibility !== false ? (
      <div key={i} style={{ display: 'flex', gap: '4px' }}>
        <div className={getClassName()} style={{ flex: '1 1 auto' }}>
          {renderSetting()}
          {errorMessage && !hideErrorMessage[i] && (
            <span className="msla-input-parameter-error" role="alert">
              {errorMessage}
            </span>
          )}
        </div>
        {!readOnly && conditionalVisibility === true ? (
          <Tooltip relationship="label" content={removeParamTooltip}>
            <Button
              appearance="subtle"
              onClick={removeParamCallback}
              icon={<ClearIcon />}
              style={{ marginTop: '30px', color: 'var(--colorBrandForeground1)', height: '32px' }}
            />
          </Tooltip>
        ) : null}
      </div>
    ) : null;
  };

  return (
    <div className="msla-setting-section-settings">
      {/* Render Required, Important, or Parameters with Values*/}
      {settings?.filter((setting) => (setting.settingProp as any).conditionalVisibility === undefined).map(renderSetting)}
      {/* Render the dropdown list of advanced parameters */}
      {allConditionalSettings.length > 0 && !readOnly ? (
        <div style={{ padding: '24px 0px 16px' }}>
          <Divider style={{ padding: '16px 0px' }} />
          <SearchableDropdownWithAddAll
            dropdownProps={{
              multiSelect: true,
              options: conditionallyInvisibleSettings.map(
                (setting): IDropdownOption => ({
                  key: (setting.settingProp as any).id,
                  text: (setting.settingProp as any).label ?? '',
                })
              ),
              placeholder: addNewParamText,
            }}
            onItemSelectionChanged={(parameterId, value) => {
              dispatch(updateParameterConditionalVisibility({ nodeId, groupId: id ?? '', parameterId, value }));
            }}
            onShowAllClick={() =>
              conditionallyInvisibleSettings.forEach((setting) =>
                dispatch(
                  updateParameterConditionalVisibility({
                    nodeId,
                    groupId: id ?? '',
                    parameterId: (setting.settingProp as any).id,
                    value: true,
                  })
                )
              )
            }
            onHideAllClick={() =>
              allConditionalSettings.forEach(
                (setting) =>
                  (setting.settingProp as any).conditionalVisibility &&
                  dispatch(
                    updateParameterConditionalVisibility({
                      nodeId,
                      groupId: id ?? '',
                      parameterId: (setting.settingProp as any).id,
                      value: false,
                    })
                  )
              )
            }
            addAllButtonText={addAllButtonText}
            addAllButtonTooltip={addAllButtonTooltip}
            addAllButtonEnabled={conditionallyInvisibleSettings.length > 0}
            removeAllButtonText={removeAllButtonText}
            removeAllButtonTooltip={removeAllButtonTooltip}
            removeAllButtonEnabled={conditionallyInvisibleSettings.length !== allConditionalSettings.length}
            label={advancedParametersLabel}
          />
        </div>
      ) : null}
      {/* Render all advanced parameters that are conditionally visible */}
      {settings?.filter((setting) => (setting.settingProp as any).conditionalVisibility === true).map(renderSetting)}
    </div>
  );
};

const matchErrorTypeToMessageBar = (errorType: ValidationErrorType): MessageBarType => {
  switch (errorType) {
    case ValidationErrorType.ERROR:
      return MessageBarType.error;
    case ValidationErrorType.WARNING:
      return MessageBarType.warning;
    case ValidationErrorType.INFO:
      return MessageBarType.info;
    default:
      return MessageBarType.info;
  }
};
