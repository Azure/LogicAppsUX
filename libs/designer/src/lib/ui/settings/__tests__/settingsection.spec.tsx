import { SettingsSection } from '../settingsection';
import type { SettingsSectionProps } from '../settingsection';
import * as ReactShallowRenderer from 'react-test-renderer/shallow';
import { describe, beforeEach, afterEach, it, expect } from 'vitest';
describe('ui/settings/settingsection', () => {
  let minimal: SettingsSectionProps;
  let renderer: ReactShallowRenderer.ShallowRenderer;

  beforeEach(() => {
    minimal = {
      id: 'this is an ID',
      title: 'Sample Setting Section',
      expanded: false,
      settings: [
        {
          settingType: 'MultiSelectSetting',
          settingProp: {
            options: [
              {
                label: 'Label 1',
                value: 'Label 1 Value',
              },
              {
                label: 'Label 2',
                value: 'Label 2 Value',
              },
              {
                label: 'Label 3',
                value: 'Label 3 Value',
              },
              {
                label: 'Label 4',
                value: 'Label 4 Value',
              },
            ],
            selections: [],
          },
          visible: true,
        },
        {
          settingType: 'SettingTextField',
          settingProp: {
            label: 'test label',
            value: 'test value',
          },
          visible: true,
        },
        {
          settingType: 'ReactiveToggle',
          settingProp: {
            textFieldValue: 'This is a test value',
            textFieldLabel: 'Test Label',
            onText: 'On',
            offText: 'Off',
          },
          visible: true,
        },
        {
          settingType: 'MultiAddExpressionEditor',
          settingProp: { onExpressionsChange: () => console.log('function not implemented') },
          visible: true,
        },
        {
          settingType: 'CustomValueSlider',
          settingProp: { minVal: 10, maxVal: 300, value: 200, sliderLabel: 'Slider', onValueChange: () => null },
          visible: true,
        },
        {
          settingType: 'MultiSelectSetting',
          settingProp: {
            options: [
              {
                label: 'Label 1',
                value: 'Label 1 Value',
              },
              {
                label: 'Label 2',
                value: 'Label 2 Value',
              },
              {
                label: 'Label 3',
                value: 'Label 3 Value',
              },
              {
                label: 'Label 4',
                value: 'Label 4 Value',
              },
            ],
            selections: [],
          },
          visible: true,
        },
      ],
      onHeaderClick: () => {
        jest.fn();
      },
    };
    renderer = ReactShallowRenderer.createRenderer();
  });

  afterEach(() => {
    renderer.unmount();
  });

  it('should construct', () => {
    const settingSection = renderer.render(<SettingsSection {...minimal} />);
    expect(settingSection).toMatchSnapshot();
  });

  it('should have child section when expanded', () => {
    const props: SettingsSectionProps = { ...minimal, expanded: true };
    renderer.render(<SettingsSection {...props} />);

    const settingSection = renderer.getRenderOutput();
    expect(settingSection.props.className).toBe('msla-setting-section');

    const settingSectionChild = settingSection.props.children;
    expect(settingSectionChild.props.className).toBe('msla-setting-section-content');
  });
});
