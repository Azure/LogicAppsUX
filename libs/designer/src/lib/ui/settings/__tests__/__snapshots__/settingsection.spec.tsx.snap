// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/settings/settingsection should construct 1`] = `
<div
  className="msla-setting-section"
>
  <div
    className="msla-setting-section-content"
  >
    <button
      className="msla-setting-section-header"
      onClick={[Function]}
    >
      <Memo(Icon)
        aria-label="Collapsed Sample Setting Section, Click to Expand"
        className="msla-setting-section-header-icon"
        iconName="ChevronRightMed"
        styles={
          Object {
            "root": Object {
              "color": "#514f4e",
              "fontSize": 14,
            },
          }
        }
      />
      <div
        className="msla-setting-section-header-text"
      >
        Sample Setting Section
      </div>
    </button>
    <React.Fragment>
      <Separator
        className="msla-setting-section-separator"
        styles={
          Object {
            "root": Object {
              "color": "#eaeaea",
            },
          }
        }
      />
    </React.Fragment>
  </div>
</div>
`;
